rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -172,9 +169,157 @@ NodeParameters::NodeParameters(
+}
+
+RCLCPP_LOCAL
+rcl_interfaces::msg::SetParametersResult
+__declare_parameter_common(
+  const std::string & name,
+  const rclcpp::ParameterValue & default_value,
+  rcl_interfaces::msg::ParameterDescriptor parameter_descriptor,
+  std::map<std::string, rclcpp::node_interfaces::ParameterInfo> & parameters_out,
+  const std::map<std::string, rclcpp::ParameterValue> & initial_values,
+  OnParametersSetCallbackType on_set_parameters_callback,
+  rcl_interfaces::msg::ParameterEvent * parameter_event_out)
+{
+  using rclcpp::node_interfaces::ParameterInfo;
+  std::map<std::string, ParameterInfo> parameter_infos {{name, ParameterInfo()}};
+  parameter_infos.at(name).descriptor = parameter_descriptor;
+
+  // Use the value from the initial_values if available, otherwise use the default.
+  const rclcpp::ParameterValue * initial_value = &default_value;
+  auto initial_value_it = initial_values.find(name);
+  if (initial_value_it != initial_values.end()) {
+    initial_value = &initial_value_it->second;
+  }
+
+  // Check with the user's callback to see if the initial value can be set.
+  std::vector<rclcpp::Parameter> parameter_wrappers {rclcpp::Parameter(name, *initial_value)};
+  // This function also takes care of default vs initial value.
+  auto result = __set_parameters_atomically_common(
+    parameter_wrappers,
+    parameter_infos,
+    on_set_parameters_callback);
+
+  // Add declared parameters to storage.
+  parameters_out[name] = parameter_infos.at(name);
+
+  // Extend the given parameter event, if valid.
+  if (parameter_event_out) {
+    parameter_event_out->new_parameters.push_back(parameter_wrappers[0].to_parameter_msg());
+  }
+
+  return result;
+}
+
+const rclcpp::ParameterValue &
+NodeParameters::declare_parameter(
+  const std::string & name,
+  const rclcpp::ParameterValue & default_value,
+  rcl_interfaces::msg::ParameterDescriptor parameter_descriptor)
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  // TODO(sloretz) parameter name validation
+  if (name.empty()) {
+    throw rclcpp::exceptions::InvalidParametersException("parameter name must not be empty");
+  }
+
+  // Error if this parameter has already been declared and is different
+  if (__lockless_has_parameter(parameters_, name)) {
+    throw rclcpp::exceptions::ParameterAlreadyDeclaredException(
+      "parameter '" + name + "' has already been declared");
+  }
+
+  rcl_interfaces::msg::ParameterEvent parameter_event;
+  auto result = __declare_parameter_common(
+    name,
+    default_value,
+    parameter_descriptor,
+    parameters_,
+    initial_parameter_values_,
+    on_parameters_set_callback_,
+    &parameter_event);
+
+  // If it failed to be set, then throw an exception.
+  if (!result.successful) {
+    throw rclcpp::exceptions::InvalidParameterValueException(
+      "parameter '" + name + "' could not be set: " + result.reason);
+  }
+
+  // Publish the event.
+  events_publisher_->publish(parameter_event);
+
+  return parameters_.at(name).value;
```

Comment:
If the type changes at all then it is invalid. I'd like to know what others think, because I'm not sure.

On the one hand, we return `const &` to private members often and `std::variant` (the closet alternative to our `ParameterValue`) allows you to get const and non-const references to storage:

https://en.cppreference.com/w/cpp/utility/variant/get

Even though, I guess it must be invalid after you do an `emplace` or `swap` on it...

Perhaps it is sufficient to document what should be obvious, which is that the reference is only valid until you change the parameter. I think it would be good to have this feature as reading the parameters is a common case (done often) and could be much faster if a reference is returned when the parameter value is large.

Reply:
Yeah, I think as long as it's documented it should be fine.

Reply:
Actually, I think I was wrong (I assumed this was the version that returned the actual value and not a `ParameterValue`). I think this reference will be valid until you undeclare it, because even if you set it, then it should be changed with the assignment operator (it's not stored as a pointer). Also, it is robust to changes when you mutate the `parameters_` map, since the reference returned by `at()` is valid even when inserting or deleting:

https://stackoverflow.com/a/4555645/671658

So long as you don't delete the parameter you're reference points to (indirectly in this case).

Reply:
I added a note about this in the docs, however.